### PR TITLE
M-02 [Oval] Attempts to Push Price to the CoinbaseOracle Will Always Fail

### PR DESCRIPTION
**Pull request by md0x**
Fixes the following issue reported in the Oval Incremental Audit:
```
The CoinbaseOracle contract serves as an oracle for price data reported and signed by
Coinbase. In order to supply such price data to the contract, users have to call the
pushPrice function which, in case of the correct data passed, updates the price data for a
given asset. In order to verify the correctness of provided data, several checks are made,
including the check for a valid kind field. However, this check reverts when kind is different
than "price". The data signed by the Coinbase oracle will always contain kind equal to
"prices", resulting in the pushPrice function always reverting, making it impossible to
provide asset prices.
Consider verifying that the provided kind value is equal to "prices".
```

Changes:
- CoinbaseOracle now takes a `string memory _dataKind` argument instead of having a hardcoded "price" value
- Add forge tests that push real prices from Coinbase Oracle API
- Small interface updates to simplify contract

### CODE REVIEW

File: src/interfaces/coinbase/IAggregatorV3SourceCoinbase.sol [hunk 1]
```diff
@@ -7,10 +7,10 @@ interface IAggregatorV3SourceCoinbase {
     function latestRoundData(string memory ticker)
         external
         view
-        returns (uint80 roundId, int256 answer, uint256 startedAt, uint256 updatedAt, uint80 answeredInRound);
```
Comment (md0x) [inline src/interfaces/coinbase/IAggregatorV3SourceCoinbase.sol:10]: Additionally, we can simplify the return values, given that they don’t match the Chainlink Aggregator interface anymore with the ticker argument.

File: src/oracles/CoinbaseOracle.sol [hunk 2]
```diff
@@ -25,31 +29,30 @@ contract CoinbaseOracle is IAggregatorV3SourceCoinbase {
     /**
      * @notice Constructor to initialize the CoinbaseOracle contract.
      * @param _decimals The number of decimals in the reported price.
+     * @param _dataKind The kind of data that the oracle will report, expected to be "prices" but open for changes.
      * @param _reporter The address of the reporter allowed to push price data.
      */
-    constructor(uint8 _decimals, address _reporter) {
+    constructor(uint8 _decimals, string memory _dataKind, address _reporter) {
```
Comment (md0x) [inline src/oracles/CoinbaseOracle.sol:35]: We now can pass a data kind in the constructor

Comment (chrismaree) [reply src/oracles/CoinbaseOracle.sol:35]: if it's always `prices` then we should not yet the deployer to define this, I think? if it should always be a hard-coded value then we should hard-code it.

File: src/oracles/CoinbaseOracle.sol [hunk 3]
```diff
@@ -58,46 +55,47 @@ contract CoinbaseOracle is IAggregatorV3SourceCoinbase {
...
         (
-            string memory kind, // e.g. "price"
+            string memory kind, // e.g. "prices"
             uint256 timestamp, // e.g. 1629350000
             string memory ticker, // e.g. "BTC"
             uint256 price // 6 decimals
         ) = abi.decode(priceData, (string, uint256, string, uint256));
 
-        require(keccak256(abi.encodePacked(kind)) == keccak256(abi.encodePacked("price")), "Invalid kind.");
+        require(keccak256(abi.encodePacked(kind)) == keccak256(abi.encodePacked("prices")), "Invalid kind.");
```
Comment (Reinis-FRP) [inline src/oracles/CoinbaseOracle.sol:88]: it might be better to use constant for KIND

File: src/oracles/CoinbaseOracle.sol [hunk 4]
```diff
@@ -8,48 +8,45 @@ import {IAggregatorV3SourceCoinbase} from "../interfaces/coinbase/IAggregatorV3S
...
-     * @param _decimals The number of decimals in the reported price.
-     * @param _reporter The address of the reporter allowed to push price data.
+     * @notice Returns the address of the reporter.
+     * @return The address of the reporter.
      */
-    constructor(uint8 _decimals, address _reporter) {
-        decimals = _decimals;
-        reporter = _reporter;
+    function reporter() public view virtual returns (address) {
+        return 0xfCEAdAFab14d46e20144F48824d0C09B1a03F2BC;
```
Comment (Reinis-FRP) [inline src/oracles/CoinbaseOracle.sol:32]: better leave as immutable. If mock needs override then can do that in mock constructor

File: src/oracles/CoinbaseOracle.sol (hunk 3 above)
Comment (Reinis-FRP) [reply src/oracles/CoinbaseOracle.sol:88]: or immutable `bytes32 KIND_HASH = keccak256(abi.encodePacked("prices");`

File: src/oracles/CoinbaseOracle.sol [hunk 5]
```diff
@@ -8,48 +8,45 @@ import {IAggregatorV3SourceCoinbase} from "../interfaces/coinbase/IAggregatorV3S
  * @notice A smart contract that serves as an oracle for price data reported by a designated reporter.
  */
 contract CoinbaseOracle is IAggregatorV3SourceCoinbase {
-    address immutable reporter;
+    uint8 public immutable decimals = 6;
```
Comment (chrismaree) [inline src/oracles/CoinbaseOracle.sol:11]: also if we make it immutable I think it should be all uppercase, no?

**Review by chrismaree — COMMENTED** [review]
looks good so far but needs a bit of refinement to be more consistent. ever immutable should be uppercase, for example.

**Review by Reinis-FRP — APPROVED** [review]

**Pull request merged**